Optimize and modernize code.

teo-tsirpanis opened the pull request:
I took a look at the code and optimized some things to better align with modern .NET's advancements, like spans and framework APIs that do what we manually used to do, and cleaned-up other stuff I encountered.

My changes are better described in the individual commit messages, feel free to ask if you have any questions or objections.

teo-tsirpanis committed:
net6 -> net6.0

teo-tsirpanis committed:
Remove two redundant dependencies.

teo-tsirpanis committed:
Stop using inline IL in F#.

It is not intended to be used from user code and we can assume that the JIT will do a good job optimizing this.

teo-tsirpanis committed:
Suppress a warning about the use of unsafe code.

Now the project builds without warnings.

teo-tsirpanis committed:
Reference System.Memory.

And target .NET Standard 2.1 in addition to 2.0 to avoid it if possible.

teo-tsirpanis committed:
Use RuntimeHelpers.AllocateTypeAssociatedMemory instead of UnmanagedMemoryPool.

It does the entire work for us.

teo-tsirpanis committed:
Use spans in IdentifierValidation.

teo-tsirpanis committed:
Add [Flags] to an enum.

teo-tsirpanis committed:
Remove an outdated comment.

OpCode is just 8 bytes long in .NET 7.

teo-tsirpanis committed:
Entirely remove PackageLicenseUrl.

teo-tsirpanis committed:
Avoid mutating strings in Text.Concat.

And clean-up the code a bit.

teo-tsirpanis committed:
Use spans in Text.HexEscape.

teo-tsirpanis committed:
Put case folding data in native memory.

We no longer pin an array of characters forever, which would survive after unloading.

teo-tsirpanis committed:
Remove NoInlining from throw helpers and make them static.

The JIT will not recognize the pattern if they have NoInlining.

teo-tsirpanis committed:
Avoid allocating the encoding preamble when we can.

teo-tsirpanis committed:
Remove optimizations in Emit.loadI4.

.NET does them by itself.

teo-tsirpanis committed:
Obsolete `Text.Is***Surrogate` and replace it with the framework method.

We don't care about Silverlight anymore, right?

teo-tsirpanis committed:
Make `IsWhitespaceHelper` a static class.

teo-tsirpanis committed:
Remove allocations in Text.CountTextElements.

stephan-tolksdorf commented:
Thank you so much for the PR!

I'd like to eventually reunite the different versions of FParsec. We'd then have to target `netstandard2.0` and `net7.0`. Do you think it would still be worthwhile to also target `netstandard2.1`?

teo-tsirpanis commented:
Yes. Users on .NET earlier than 7 would get the .NET Standard 2.1 edition which does not reference compatibility packages for spans and reflection emit.

teo-tsirpanis commented on FParsecCS/Text.cs:
```diff
@@ -395,11 +392,11 @@ public static int CountTextElements(string str)
 #else
 
-internal unsafe struct IsWhitespaceHelper {
+internal unsafe static class IsWhitespaceHelper {
```
How about I remove this entire custom `IsWhitespace` implementation? .NET does something similar.

https://github.com/dotnet/runtime/blob/ec26662ac9f1e89b2556d81fa311f91ab3abd492/src/libraries/System.Private.CoreLib/src/System/Globalization/CharUnicodeInfo.cs#L182-L191

stephan-tolksdorf replied:
Yeah, the performance is probably similar nowadays, though it may be worth double checking.

teo-tsirpanis replied:
😳

|  Method |     Mean |   Error |  StdDev |
|-------- |---------:|--------:|--------:|
|  DotNet | 325.0 μs | 3.88 μs | 3.63 μs |
| FParsec | 225.8 μs | 4.26 μs | 3.78 μs |

```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using FParsec;

BenchmarkRunner.Run<IsWhitespaceBenchmark>(args: args);

public class IsWhitespaceBenchmark
{
    [Benchmark]
    public int DotNet()
    {
        int count = 0;
        for (char c = '\0'; c != char.MaxValue; c++)
            if (char.IsWhiteSpace(c))
                count++;
        return count;
    }
    
    [Benchmark]
    public int FParsec()
    {
        int count = 0;
        for (char c = '\0'; c != char.MaxValue; c++)
            if (Text.IsWhitespace(c))
                count++;
        return count;
    }
}
```

stephan-tolksdorf replied:
Thanks for measuring! Then maybe let's keep the custom implementation until it becomes a significant hassle to maintain.

stephan-tolksdorf commented:
Excuse my ignorance, what is the downside of referencing the compatibility packages?

teo-tsirpanis commented:
More package references. 😛

stephan-tolksdorf commented:
I just wondered how to weigh this against increasing the Nuget package size. It seems most other packages target netstandard2.0 + netX.0

teo-tsirpanis commented:
`FParsec 2.0.0-beta` weighs 186 KB. `System.Memory` weighs 204 KB. The low-trust edition package I built weighs 347 KB, which is less than the sum of the two packages above. Tbh I never thought of the increasing package size but we are talking about low numbers below half a megabyte, so I wouldn't bother. Whatever you choose will not have a runtime impact.

stephan-tolksdorf commented on FParsecCS/Text.cs:
```diff
@@ -204,7 +204,11 @@ public static class Text {
 /// <summary>Returns the given string with all occurrences of "\r\n" and "\r" replaced
 /// by "\n". If the argument is null, null is returned.</summary>
+#if NET
+public static string NormalizeNewlines(string str) => str?.ReplaceLineEndings("\n");
```
That's a behaviour change, since it replaces all Unicode newlines, not just the three FParsec normally handles. Is that intentional?

teo-tsirpanis replied:
I was misled by the documentation above. Thanks for pointing out. Reverted.

stephan-tolksdorf commented on FParsecCS/Text.cs:
```diff
@@ -363,11 +363,12 @@ public static int CountTextElements(string str)
 
-// Apparently System.Char.Is(High|Low)Surrogate is not safe for consumption by Silverlight developers
-
-public static bool IsSurrogate(char ch)     { return (ch & 0xF800) == 0xD800; }
```
The definitions here are slightly simpler than the system ones, but that shouldn't make a difference on modern processors, and hopefully not for the JIT inliner/optimizer either.

stephan-tolksdorf commented on FParsec/Emit.fs:
```diff
@@ -22,16 +22,23 @@ open FParsec.Range
+let private createTypeBuilderSyncRoot = obj()
+
+let getModuleBuilder() =
+    match moduleBuilder with
```
I don't think this access is safe. In C# you'd need to make the `moduleBuilder` variable volatile in order to ensure the variable is loaded with acquire semantics. This also seems to match the [draft memory model documentation](https://github.com/dotnet/runtime/pull/75790/files).

teo-tsirpanis replied:
[There is no `volatile` in the memory document example.](https://github.com/dotnet/runtime/pull/75790/files#diff-34c8358c0bb94c612b0d6d189eb597202919dc4785920041081583b87981287eR208-R234) `lock` places its own barriers.

stephan-tolksdorf replied:
Maybe that's an oversight or error. I've haven't seen an argument why this is safe in general. The read outside the lock doesn't use acquire semantics and isn't synchronized with the lock operations.

teo-tsirpanis replied:
I eventually just used `lazy`. 😁

stephan-tolksdorf commented on FParsecCS/Text.cs:
```diff
@@ -364,13 +364,17 @@ public static class Text {
 /// <summary>Returns System.Globalization.StringInfo(str).LengthInTextElements</summary>
 public static int CountTextElements(string str)
 {
+#if NET
```
Please later squash this into the previous commit.

teo-tsirpanis replied:
Done.

stephan-tolksdorf reviewed (commented):
These are great contributions! Almost ready to be merged! 😃

teo-tsirpanis committed:
Use spans in HexFloat.

teo-tsirpanis commented on FParsecCS/Buffer.cs:
```diff
@@ -23,188 +25,38 @@ public static class Buffer {
 
 #if !LOW_TRUST
 // used by StaticMapping.createStaticStringMapping
 public static unsafe bool Equals(uint* ptr1, uint* ptr2, uint length) {
```
I could make this use `MemoryExtensions.SequenceEqual` if the length was an `int`…

stephan-tolksdorf replied:
Should be safe for string comparisons. But you'd have to touch `StaticMapping.fs` 😁

teo-tsirpanis replied:
Turns out the length is a compile-time constant and always signed. Done.

teo-tsirpanis commented on FParsecCS/CaseFoldTable.cs:
```diff
@@ -46,13 +42,7 @@ internal static class CaseFoldTable {
-        // so pinning it is a no-op and the GC is not affected.
-        FoldedCharsHandle = GCHandle.Alloc(FoldedCharsArray, GCHandleType.Pinned);
-        return (char*)FoldedCharsHandle.AddrOfPinnedObject();
+        return chars;
```
[Modern .NET supports unloadability but it works cooperatively.](https://learn.microsoft.com/en-us/dotnet/standard/assembly/unloadability) To be able to unload an `AssemblyLoadContext`, no direct or indirect strong references to it must exist. `AllocateTypeAssociatedMemory` says "if this type gets unloaded, release this memory too".

stephan-tolksdorf replied:
You're right, I looked at the wrong `AllocateTypeAssociatedMemory`. Thank you!

teo-tsirpanis committed:
Simplify locking when creating type builders.

We just use Lazy to lock the assembly builder creation. .NET has its own locks for the method builders.

teo-tsirpanis committed:
Always use the full `ldloc` and `stloc` opcodes.

They are brittle if we change the order of locals and ILGenerator picks the most compact opcode.

teo-tsirpanis committed:
Associate the native memory of the emitted code with the dynamic assembly.

We previously associated the memory with the FParsec assembly.
Because the dynamic assembly does not depend on FParsec, it is possible for the former to outlive the latter, which would cause the memory to be swept under its feet.
The memory is now associated with a dummy type we create in the dynamic assembly.
Since the dynamic assembly is not created with RunAndCollect we could have just allocated the memory with a malloc, but it can still be unloaded if it belongs in an unloadable AssemblyLoadContext.
We could have also used `TypeBuilder.DefineInitializedData` but it would need some refactoring or more allocations, and there are few benefits.

teo-tsirpanis committed:
Use framework methods in Buffer.cs.

teo-tsirpanis committed:
Use `ilsigptr` instead of `MakePointerType`.

teo-tsirpanis commented on FParsecCS/HexFloat.cs:
```diff
@@ -32,45 +32,23 @@ public static class HexFloat {
     const int expBits = 11;  // bits for biased exponent
     const int maxBits = 53;  // significant bits (including implicit bit)
-#if LOW_TRUST
-    const int maxChars = 24; // "-0x1.fffffffffffffp-1022"
```
Done.

teo-tsirpanis commented on FParsecCS/HexFloat.cs:
```diff
@@ -98,7 +98,11 @@ public static class HexFloat {
+    uint xn = BitConverter.ToUInt32(BitConverter.GetBytes(x), 0);
+    #else
     uint xn = (uint)BitConverter.SingleToInt32Bits(x);
+    #endif
```
Done.

stephan-tolksdorf reviewed (changes requested):
Could you maybe add a small summary of your work in doc/src/changelog.txt for v2.0?

teo-tsirpanis committed:
Optimize `Buffer.Equals`.

Turns out the length can be a signed integer which means we can use SequenceEqual.

teo-tsirpanis committed:
Change `#if !X #else` to `#if !X`.

teo-tsirpanis committed:
Stop using TypeInfo. All the frameworks we target support them.

teo-tsirpanis committed:
Limit the use of ref cells and their deprecated operators.

teo-tsirpanis committed:
Replace `x.[i]` with `x[i]`.

teo-tsirpanis committed:
Use spans in Text, just like IdentifierValidation.

teo-tsirpanis committed:
Make FastGenericEqualityERComparer<T>.Instance readonly.

It might help the JIT to devirtualize.

teo-tsirpanis committed:
Remove PCL.

teo-tsirpanis committed:
Update the changelog.

teo-tsirpanis commented:
Feedback addressed; I also cleaned-up more stuff.

Can I remove `UNALIGNED_READS`, making it always on?

stephan-tolksdorf commented:
Thank you this is looking good! However, I won't get to fully review it and publish a new FParsec release until this weekend.

I think I saw code in the runtime assuming unaligned reads work ok. Do you know whether this is documented somewhere?

teo-tsirpanis commented:
I can wait, but my recommendation is to not merge this until #94 and #96 get resolved.

Regarding unaligned reads, if I remember correctly the only platform they pose a problem on is the rare Arm32. The correct way to do unaligned reads is through the `unaligned` IL prefix, which is exposed in the form of methods like [`Unsafe.ReadUnaligned`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.readunaligned).

teo-tsirpanis committed:
Fix a compiler error.

teo-tsirpanis commented:
CI is green. 🚀

stephan-tolksdorf commented:
Great work! Thank you!
